kernel/posix/pthread.c

```diff
@@ -0,0 +1,537 @@
+	pthread_mutex_lock(&pthread->cancel_lock);
+	pthread->cancel_state = state;
+
+	if (state == PTHREAD_CANCEL_DISABLE) {
+		_thread_essential_set();
+	} else {
+		_thread_essential_clear();
+	}
+
+	pthread_mutex_unlock(&pthread->cancel_lock);
+
+	if (state == PTHREAD_CANCEL_ENABLE && pthread->cancel_pending) {
+		pthread_exit((void *)PTHREAD_CANCELED);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Cancel execution of a thread.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_cancel(pthread_t pthread)
+{
+	struct posix_thread *thread = (struct posix_thread *) pthread;
+
+	if (thread == NULL || thread->state == PTHREAD_TERMINATED) {
+		return ESRCH;
+	}
+
+	pthread_mutex_lock(&thread->cancel_lock);
+	thread->cancel_pending = 1;
+
+	if (thread->cancel_state == PTHREAD_CANCEL_ENABLE) {
+		if (thread->state == PTHREAD_DETACHED) {
+			thread->state = PTHREAD_TERMINATED;
+			pthread_mutex_unlock(&thread->cancel_lock);
+		} else {
+			thread->retval = PTHREAD_CANCELED;
+			pthread_mutex_unlock(&thread->cancel_lock);
+			pthread_mutex_lock(&thread->state_lock);
+			thread->state = PTHREAD_EXITED;
+			pthread_cond_broadcast(&thread->state_cond);
+			pthread_mutex_unlock(&thread->state_lock);
+		}
+
+		k_thread_abort((k_tid_t) thread);
+	} else {
+		pthread_mutex_unlock(&thread->cancel_lock);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Set thread scheduling policy and parameters.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_setschedparam(pthread_t pthread, int policy,
+			  const struct sched_param *param)
+{
+	k_tid_t *thread = (k_tid_t *)pthread;
+	int new_prio, cur_prio;
+
+	if (thread == NULL) {
+		return ESRCH;
+	}
+
+	if (policy != SCHED_RR && policy != SCHED_FIFO) {
+		return EINVAL;
+	}
+
+	new_prio = posix_to_zephyr_priority(param->priority, policy);
+	cur_prio = k_thread_priority_get(*thread);
+
+	if (is_posix_prio_valid(new_prio, policy) == false ||
+	    new_prio == cur_prio) {
+		return EINVAL;
+	}
+
+	k_thread_priority_set(*thread, new_prio);
+
+	return 0;
+}
+
+/**
+ * @brief Initialise threads attribute object
+ *
+ * See IEEE 1003.1
+ */
+int pthread_attr_init(pthread_attr_t *attr)
+{
+
+	if (attr->initialized == true) {
+		return EBUSY;
+	}
+
+	*attr = init_pthread_attrs;
+
+	return 0;
+}
+
+/**
+ * @brief Get thread scheduling policy and parameters
+ *
+ * See IEEE 1003.1
+ */
+int pthread_getschedparam(pthread_t pthread, int *policy,
+			  struct sched_param *param)
+{
+	k_tid_t thread = (k_tid_t)pthread;
+	u32_t  priority = k_thread_priority_get(thread);
+
+	param->priority = zephyr_to_posix_priority(priority, policy);
+	return 0;
+}
+
+/**
+ * @brief Terminate calling thread.
+ *
+ * See IEEE 1003.1
+ */
+void pthread_exit(void *retval)
+{
+	struct posix_thread *self = (struct posix_thread *)pthread_self();
+
+	/* Make a thread as cancelable before exiting */
+	pthread_mutex_lock(&self->cancel_lock);
+	if (self->cancel_state == PTHREAD_CANCEL_DISABLE) {
+		self->cancel_state = PTHREAD_CANCEL_ENABLE;
+		_thread_essential_clear();
```

Comment:
This function is called by the pthread wrapper, so it might be a good idea to mention that pthread won't work at all in user mode at the moment.

Reply:
I was thinking of removing this reference to  _thread_essential_clear and _thread_essential_set and handle through PTHREAD_CANCEL_ENABLE entirely.

Yes as of now the POSIX implementation will not work with user mode and there are other candidates too which are not user mode compliant. Example: mempool too is not user mode compliant.